Add torch.Assert, which is symbolically traceable

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45188 Add torch.Assert, which is symbolically traceable**

Summary:

This is a symbolically traceable alternative to Python's `assert`.
It should be useful to allow people who want to use FX to also
be able to assert things.

A bunch of TODO(before) land are inline - would love thoughts
on where is the best place for this code to live, and what this
function should be called (since `assert` is reserved).  TF uses
`Assert`, which seems reasonable.

Test Plan:

```
python test/test_utils.py TestAssert
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23861567](https://our.internmc.facebook.com/intern/diff/D23861567)

vkuzo committed:
Add torch.do_assert, which is symbolically traceable

Summary:

This is a symbolically traceable alternative to Python's `assert`.
It should be useful to allow people who want to use FX to also
be able to assert things.

A bunch of TODO(before) land are inline - would love thoughts
on where is the best place for this code to live, and what this
function should be called (since `assert` is reserved).

Test Plan:

```
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

yf225 commented on test/test_fx.py:
```diff
@@ -600,6 +600,22 @@ def test_construct_root_dict(self):
+
+        class AssertsTensorShape(torch.nn.Module):
+            def forward(self, x):
+                torch.do_assert(x.shape[1] > 4, message)
```
Is there a way to make `torch.do_assert` work in non-symbolic-tracing mode as well? What we want to avoid is having to branch based on whether we are in symbolic tracing mode:
```python
if is_symbolic_tracing():
  torch.do_assert(x.shape[1] > 4, message)
else:
  assert x.shape[1] > 4, message
```
As we use assertions extensively in DPER3 code, having to add branches similar to this will make the code much harder to maintain.

vkuzo replied:
yes, it works both traced and non-traced without needing to branch.  I'll add a test case.

vkuzo committed:
Update on "Add torch.do_assert, which is symbolically traceable"


Summary:

This is a symbolically traceable alternative to Python's `assert`.
It should be useful to allow people who want to use FX to also
be able to assert things.

A bunch of TODO(before) land are inline - would love thoughts
on where is the best place for this code to live, and what this
function should be called (since `assert` is reserved).

Test Plan:

```
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23861567](https://our.internmc.facebook.com/intern/diff/D23861567)

[ghstack-poisoned]

ezyang commented on torch/csrc/Module.cpp:
```diff
@@ -614,6 +628,7 @@ static PyMethodDef TorchMethods[] = {
   {"_is_xnnpack_enabled", (PyCFunction)THPModule_isEnabledXNNPACK, METH_NOARGS, nullptr},
   {"_is_torch_function_enabled", (PyCFunction)THPModule_isEnabledTorchFunction, METH_NOARGS, nullptr},
   {"_disabled_torch_function_impl", (PyCFunction)THPModule_disable_torch_function, METH_VARARGS, nullptr},
+  {"Assert", (PyCFunction)THPModule_Assert, METH_VARARGS, nullptr},
```
why... capitalized?

ezyang replied:
oh because assert is a reserved keyword

vkuzo replied:
`assert` is reserved in Python, just keeping the naming consistent throughout.  There is a TODO(before land) to align on best actual name to use.

ezyang commented:
Is there any point in having the C binding code? I'm not sure what it is doing that couldn't just have been done in Python land.

vkuzo commented:
> Is there any point in having the C binding code? I'm not sure what it is doing that couldn't just have been done in Python land.

ah, good point, will fix

vkuzo committed:
Update on "Add torch.Assert, which is symbolically traceable"


Summary:

This is a symbolically traceable alternative to Python's `assert`.
It should be useful to allow people who want to use FX to also
be able to assert things.

A bunch of TODO(before) land are inline - would love thoughts
on where is the best place for this code to live, and what this
function should be called (since `assert` is reserved).  TF uses
`Assert`, which seems reasonable.

Test Plan:

```
python test/test_utils.py TestAssert
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23861567](https://our.internmc.facebook.com/intern/diff/D23861567)

[ghstack-poisoned]

dzhulgakov reviewed (commented):
Here's a hacky idea - we can detect the assert statement itself: https://colab.research.google.com/drive/1nmoCr5ZBjSc1t5StsZoIMgmcXOma1BQi?usp=sharing

It's pretty magical, but we already have precedent for keywords unpacking here https://github.com/pytorch/pytorch/blob/master/torch/fx/proxy.py#L97 and it works just fine. Why not add one more? :)

yf225 commented:
> Here's a hacky idea - we can detect the assert statement itself: https://colab.research.google.com/drive/1nmoCr5ZBjSc1t5StsZoIMgmcXOma1BQi?usp=sharing
> 
> It's pretty magical, but we already have precedent for keywords unpacking here https://github.com/pytorch/pytorch/blob/master/torch/fx/proxy.py#L97 and it works just fine. Why not add one more? :)

Agreed with Dima - it would be really awesome to hide FX-specific requirements and make the native `assert` just work for FX. Otherwise I am a bit worried that this paves the way for more PyTorch-specific control flow statements down the road..

vkuzo commented:
> Here's a hacky idea - we can detect the assert statement itself: https://colab.research.google.com/drive/1nmoCr5ZBjSc1t5StsZoIMgmcXOma1BQi?usp=sharing
> 
> It's pretty magical, but we already have precedent for keywords unpacking here https://github.com/pytorch/pytorch/blob/master/torch/fx/proxy.py#L97 and it works just fine. Why not add one more? :)

ah, nice.  Pending any concerns from core FX team, that sgtm.  Thx for the example code!

jamesr66a commented:
@dzhulgakov that seems too messy and reaches too far into the implementation details of cpython. My suggestion on the original issue (https://github.com/pytorch/pytorch/issues/44849#issuecomment-696255749) was to do an AST-based rewrite to swap out [Assert](https://greentreesnakes.readthedocs.io/en/latest/nodes.html#Assert) nodes with a callsite to this op, for example. I think this is a more robust solution (at least I believe the Python AST will be relatively stable, owing to our experience with the TorchScript frontend). Just need someone to work on it

jamesr66a commented:
We could maybe cannibalize some of the code in the TorchScript frontend https://github.com/pytorch/pytorch/blob/master/torch/jit/frontend.py

jamesr66a reviewed (approved):
![image](https://user-images.githubusercontent.com/4685384/94212394-ef2e0780-fe88-11ea-8704-d25e49ed211c.png)

dzhulgakov commented:
Agree on AST rewrite, we'd need it eventually. We will basically get full autograph sooner or later :)

vkuzo committed:
Update on "Add torch.Assert, which is symbolically traceable"


Summary:

This is a symbolically traceable alternative to Python's `assert`.
It should be useful to allow people who want to use FX to also
be able to assert things.

A bunch of TODO(before) land are inline - would love thoughts
on where is the best place for this code to live, and what this
function should be called (since `assert` is reserved).  TF uses
`Assert`, which seems reasonable.

Test Plan:

```
python test/test_utils.py TestAssert
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23861567](https://our.internmc.facebook.com/intern/diff/D23861567)

[ghstack-poisoned]
